HW2/P4/P4.txt

```diff
@@ -0,0 +1,20 @@
+Experiment Results: 
+
+4 threads
+1.47259497643 seconds for 10 filter passes.
+1.50553798676 seconds for 10 filter passes.
+1.48426318169 seconds for 10 filter passes.
+
+2 threads
+1.63110995293 seconds for 10 filter passes.
+1.6989338398 seconds for 10 filter passes.
+1.72469592094 seconds for 10 filter passes.
+
+1 thread 
+3.3609380722 seconds for 10 filter passes.
+3.37282991409 seconds for 10 filter passes.
+3.37703895569 seconds for 10 filter passes.
+
+Above we have the runtimes for each of the configurations (1, 2, and 4 threads) in 3 runs. We see that increasing from 1 to 2 threads almost achieves the optimal speedup of 2x, but further increasing from  2 to 4 threads does not have as large of an impact on the runtime (it does go down slightly). This might me because the overhead with 4 threads begins to outweigh the cost of parralelism, while the overhead with 2 threads is much less than the computation time. Note that the communication overhead will scale with the number of threads (we need more events to keep track of). 
```

Comment:
This is similar to how I reasoned. One other consideration to make when you move from 2 to 4 threads is that, almost always, one of your threads is typically tasked with OS and I/O system processes. So you're not totally utilizing a full 4 threads. That said, you are right that there is added logistics overhead when you move to additional threads that can cause hang ups, especially with the event queuing method you implemented.